### PR TITLE
[doc] Update the doc for differentiable programming

### PR DESCRIPTION
**Pull request by erizmr**
The PR focuses on refactoring of the `kernel simplicity rule` (KSR) part due to that the KSR has been removed recently by #4041 #3897  .

### CODE REVIEW

File: docs/lang/articles/advanced/differentiable_programming.md [hunk 1]
```diff
@@ -169,6 +169,10 @@ for examples on using autodiff-based force evaluation MPM and FEM.
 As mentioned above, `ti.Tape()` can only track a 0D field as the output variable.
```
Comment (neozhaoliang) [inline docs/lang/articles/advanced/differentiable_programming.md:169]: shall we use a few sentences to describe what is differential programming and how it works at beginning of the doc?
Just mentioning it's useful in scientific computations and AI does not offer much information about what it is.

**Comment by neozhaoliang** [comment]
Some suggestions:

1. Shall we shortly describe what is differential programming and how it works at the beginning of the doc? Just mentioning it's useful in scientific computations and AI does not offer much information about the whole picture. Can we use a figure to visualize how autodiff works?

2. ```Unlike tools such as TensorFlow where immutable output buffers are generated, the imperative programming paradigm adopted in Taichi allows programmers to freely modify global fields.``` Could you explain more on the terms "immutable" and "imperative"? Or we can just avoid using these terms and just tell the readers which scenarios our autodiff does not cover?

3. Could you add a paragraph to explain how autodiff are implemented internally in pythorch/tensorflow/taichi? Even a sketch of the picture would be very informative.

File: docs/lang/articles/advanced/differentiable_programming.md [hunk 2]
```diff
@@ -296,41 +302,21 @@ func_break_rule_2.grad()
 assert x.grad[1] == 4.0
 assert x.grad[2] == 3.0
 ```
-### Kernel Simplicity Rule
 
-:::note Kernel Simplicity Rule
-Kernel body must consist of multiple simply nested for-loops. For example, each for-loop can either contain exactly one (nested) for-loop (and no other statements), or a group of statements without loops.
+### Avoid mixed usage of parallel for-loop and non-for statements
+Mixed usage of parallel for-loops and non-for statements are not supported in the autodiff system.
```
Comment (ailzhang) [inline docs/lang/articles/advanced/differentiable_programming.md:307]: Maybe also mention that using mixed parallel for-loop and non-for stmts will trigger an error? (so that users know they don't get silent erros :P )

File: docs/lang/articles/advanced/differentiable_programming.md [hunk 3]
```diff
@@ -169,6 +169,10 @@ for examples on using autodiff-based force evaluation MPM and FEM.
 As mentioned above, `ti.Tape()` can only track a 0D field as the output variable.
 If there're multiple output variables that you want to back-propagate
 gradients to inputs, `kernel.grad()` should be used instead of `ti.Tape()`.
+Different from using `ti.Tape()`, you need to set the `grad` of the output variables themselves to `1` manually
```
Comment (ailzhang) [inline docs/lang/articles/advanced/differentiable_programming.md:172]: It's not limited to `1` I guess?

**Review by ailzhang — APPROVED** [review]
LGTM!

File: docs/lang/articles/advanced/differentiable_programming.md (hunk 3 above)
Comment (erizmr) [reply docs/lang/articles/advanced/differentiable_programming.md:172]: Sorry I failed to come up with cases that require non-1 grad for the output variables. Could you please give some examples? :P

**Comment by erizmr** [comment]
> Some suggestions:
> 
> 1. Shall we shortly describe what is differential programming and how it works at the beginning of the doc? Just mentioning it's useful in scientific computations and AI does not offer much information about the whole picture. Adding some figures would be very helpful.
> 2. `Unlike tools such as TensorFlow where immutable output buffers are generated, the imperative programming paradigm adopted in Taichi allows programmers to freely modify global fields.` Could you explain more on the terms "immutable" and "imperative"? Or we can avoid using these terms and just tell the readers which scenarios our autodiff does not cover?
> 3. Could you add a paragraph to explain how autodiff are implemented internally in pythorch/tensorflow/taichi? Even a sketch of the picture would be very informative.

Thanks for the suggestions! 
1. I can try to give a brief introduction to differential programming. As for the figures, do you mean the figures of some applications or a overview of the differential programming paradigm?
2/3. "imperative programming" here is used compared to "functional programming" who can not change the state of the program (such as global fields). I think it is a good idea to explain the difference between pytorch/tensorflow/taichi, which may also help sort the concern about "imperative" and "immutable".

**Comment by neozhaoliang** [comment]
@erizmr emm, by figures I mean something like a flow chart shows how autodiff computes the gradients of the loss in the examples you mentioned?

**Review by Vissidarte-Herman — APPROVED** [review]

**Comment by erizmr** [comment]
The content related to `introduction to differential programming` and `difference between pytorch/tensorflow/taichi` requires more effort and polishing, which is planned to be added in the following PRs. The autodiff system is now under actively development, there are also new content such as demos, features to be added into the doc. :)
